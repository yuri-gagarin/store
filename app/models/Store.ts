import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;