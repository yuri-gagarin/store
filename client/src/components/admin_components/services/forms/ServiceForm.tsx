import React, { useState, useEffect, useRef }  from "react";
import { Button, Form, TextArea } from "semantic-ui-react";

export type FormState = {
  name: string;
  description: string;
  price: string;
}

interface Props {
  name: string;
  description: string;
  price: string;
  newForm: boolean;
  handleCreateService(data: FormState): void;
  handleUpdateService(data: FormState): void;
}


const ServiceForm: React.FC<Props> = ({ name, description, price, newForm, handleCreateService, handleUpdateService }): JSX.Element => {
  // local form state and ref //
  const [ formState, setFormState ] = useState<FormState>({ name, description, price });
  const serviceFormRef = useRef<HTMLDivElement>(document.createElement("div"));
  // input change listeners //
  const handleNameChange = (e: React.ChangeEvent<HTMLInputElement>):void => {
    setFormState({
      ...formState,
      name: e.target.value
    });
  };
  const handleDescriptionChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      description: e.target.value
    });
  };
  const handlePriceChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setFormState({
      ...formState,
      price: e.target.value
    });
  };
  // submit listener //
  const handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (newForm) {
      handleCreateService(formState);
    } else {
      handleUpdateService(formState);
    }
  };
  // lifecycle hooks //
  useEffect(() => {
    if (serviceFormRef.current) {
      const elem = serviceFormRef.current.getBoundingClientRect();
      window.scrollTo({
        top: elem.bottom,
        behavior: "smooth"
      })
    }
  }, [serviceFormRef]);
  
  return (
    <div className="serviceFormDiv" ref={serviceFormRef}>
      <Form id="serviceForm">
        <Form.Field>
          <label>Service name</label>
          <input 
            id="adminServiceFormNameInput"
            onChange={handleNameChange} 
            placeholder="Service name here ..." 
            value={formState.name}
          />
        </Form.Field>
        <Form.Field>
          <label>Service price</label>
          <input 
            id="adminServiceFormPriceInput"
            onChange={handlePriceChange} 
            placeholder="Service price here..." 
            value={formState.price}
          />
        </Form.Field>
        <Form.Field
          id='adminServiceFormDescInput'
          control={TextArea}
          label='Store Description'
          onChange={handleDescriptionChange}
          placeholder='Service description here...'
          value={formState.description}
         />
         {
           newForm 
            ? <Button id="adminServiceFormCreate" type='submit' onClick={handleSubmit} content= "Create  New Service" />
            : <Button id="adminServiceFormUpdate" type='submit' onClick={handleSubmit} content= "Update Service" />
         }  
      </Form>
    </div>
  );
};

export default ServiceForm;