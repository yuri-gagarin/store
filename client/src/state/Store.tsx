import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

