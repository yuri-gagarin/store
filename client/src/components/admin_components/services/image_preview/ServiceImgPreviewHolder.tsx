import React from "react";
// additional components //
import ServiceImgThumb from "./ServiceImgThumb";
// css  imports //
import "./css/serviceImgPreviewHolder.css";
// state //
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// helpers
import { deleteServiceImage } from "../actions/APIServiceActions";

// holder for image preview thumbnails //
interface Props {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>
}
const ServiceImgPreviewHolder: React.FC<Props> = ({ state, dispatch }): JSX.Element => {
  const { images } = state.serviceState.currentServiceData;

  const handleDelete = (_id: string) => {
    deleteServiceImage(_id, state, dispatch);
  };

  if (images.length > 0) {
    return (
      <div id="serviceImgPreviewHolder">
        <div className="serviceImgHolderTitle">Uploaded Images</div>
        {
          images.map((image) => {
            return (
              <ServiceImgThumb 
                key={image._id} 
                _id={image._id} 
                url={image.url} 
                handleDelete={handleDelete}
              />
            );
          })
        }
      </div>
    );
  } else {
    return (
      <div id="serviceImgPreviewNone">
        <h3>No images uploaded... you should upload some... </h3>
      </div>
    );
  }
};

export default ServiceImgPreviewHolder;


