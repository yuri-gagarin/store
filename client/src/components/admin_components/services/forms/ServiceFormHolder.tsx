import React, { useState, useEffect, useContext } from "react";
import { Button, Grid } from "semantic-ui-react";
// client routing //
import { withRouter, RouteComponentProps } from "react-router-dom";
// css imports //
import "./css/serviceFormHolder.css";
// additional components //
import ServiceForm from "./ServiceForm";
import ServiceImgPreviewHolder from "../image_preview/ServiceImgPreviewHolder";
import ServiceImgUplForm from "./ServiceImgUplForm";
import LoadingBar from "../../miscelaneous/LoadingBar";
// state //
import { Store } from "../../../../state/Store";
// api and ui actions //
import { createService, editService } from "../actions/APIServiceActions";
import { openServiceForm, closeServiceForm } from "../actions/UIServiceActions";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";
import { checkSetValues } from "../../../helpers/validationHelpers";
// types 
import { FormState } from "./ServiceForm";
interface Props extends RouteComponentProps {
  
}

type ServiceData = {
  name: string;
  description: string;
  price: string;
  images: IServiceImgData[];
}

const ServiceFormHolder: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, currentServiceData, serviceFormOpen, error } = state.serviceState;
  const { _id : serviceId, name, description, price, createdAt, editedAt } = currentServiceData;
  // local component state //
  const [ newForm, setNewForm ] = useState<boolean>(true);

  // ServiceForm toggle //
  const toggleServiceForm = () => {
    serviceFormOpen ? closeServiceForm(dispatch) : openServiceForm(dispatch);
  };
  // API handlers CREATE - EDIT //
  const handleCreateService = ({ name, price, description }: FormState): void => {
    const serviceData: ServiceData = {
      name,
      description,
      price,
      images: currentServiceData.images
    };

    createService(serviceData, dispatch)
      .then((_) => {
        // service created //
        closeServiceForm(dispatch);
      })
      .catch((_) => {
        // handle an error here possibly show details //
      });
  };
  const handleUpdateService = ({ name, description, price }: FormState): void => {
    const serviceParams: ServiceData = {
      name, description, price, images: currentServiceData.images
    };

    editService(currentServiceData._id, serviceParams, dispatch, state)
      .then((_) => {
        closeServiceForm(dispatch);
      })
      .catch((_) => {
        // handle an error , maybe show modal in thhe future //
      });
  };
  // lifecyle hooks //
  useEffect(() => {
    if (!serviceFormOpen) {
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  }, [ serviceFormOpen ]);
  // set correct data if a service is loaded //
  useEffect(() => {
      checkSetValues(currentServiceData) ? setNewForm(false) : setNewForm(true);
  }, [ currentServiceData ]);
  // component return //
  return (
    <div id="serviceFormHolder">
      { loading ? <LoadingBar /> : null }
      {
        !newForm ?
          <React.Fragment>
            <Grid.Row id="serviceFormHolderDetails">
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <h1>Details</h1>
                <div className="serviceFormHolderDetailsName">
                  <h3>Service name:</h3>
                  <p>{name}</p>
                </div>
                <div className="serviceFormHolderDetailsPrice">
                  <h3>Service price:</h3>
                  <p>{price}</p>
                </div>
                <div className="serviceFormHolderDetailsDesc">
                  <h3>Service description:</h3>
                  <p>{description}</p>
                </div>
                <div className="serviceFormHolderTimestamps">
                  <span>Created At: <strong>{ConvertDate.international(createdAt)}</strong></span>
                  <span>Edited At: <strong>{ConvertDate.international(editedAt)}</strong></span>
                </div>
              </Grid.Column>
            </Grid.Row>
            <Grid.Row>
              <Grid.Column mobile={16} tablet={14} computer={14}>
                <ServiceImgPreviewHolder state={state} dispatch={dispatch} />
              </Grid.Column>
            </Grid.Row>
            <ServiceImgUplForm />
          </React.Fragment>
          : null
      }
      <Grid.Row>
        <Grid.Column mobile={16} tablet={15} computer={14}>
          <Button 
            id="serviceFormToggleBtn" 
            onClick={toggleServiceForm} 
            content={ !serviceFormOpen ? "Open Form" : "Close Form"}
          />
          {
            serviceFormOpen ? 
              <ServiceForm 
                name={name} 
                description={description} 
                price={price}
                newForm={newForm}
                handleCreateService={handleCreateService}
                handleUpdateService={handleUpdateService}
              /> 
            : null
          }
         
        </Grid.Column>
      </Grid.Row>
    </div>
  );
};

// export for testing withour router //
export { ServiceFormHolder };
// default export //
export default withRouter(ServiceFormHolder);

