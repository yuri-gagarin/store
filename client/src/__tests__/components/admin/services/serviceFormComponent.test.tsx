import React from "react";
import { mount, ReactWrapper } from "enzyme";
import { Button } from "semantic-ui-react";
// component imports //
import ServiceFormHolder from "../../../../components/admin_components/services/forms/ServiceFormHolder";
import ServiceForm from "../../../../components/admin_components/services/forms/ServiceForm";
import ServiceImageUplForm from "../../../../components/admin_components/services/forms/ServiceImgUplForm";
import ServiceImgPreviewHolder from "../../../../components/admin_components/services/image_preview/ServiceImgPreviewHolder";
import ServiceImgPreviewThumb from "../../../../components/admin_components/services/image_preview/ServiceImgThumb";
// helpers //
import { generateCleanState } from "../../../../test_helpers/miscHelpers";
import { setMockServiceState } from "../../../../test_helpers/serviceHelpers";
import { IGlobalAppState } from "../../../../state/Store";

describe("ServiceFormHolder Component tests", () => {
  let wrapper: ReactWrapper; 
  
  describe("Default Form Holder state",  () => {

    beforeAll(() => {
      window.scrollTo = jest.fn();
      wrapper = mount<{}, typeof ServiceFormHolder>(<ServiceFormHolder state={generateCleanState()} dispatch={jest.fn<React.Dispatch<ServiceAction>, []>()}/>)
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.html()).toBeDefined();
    });
    it("Form Should be closed by default", () => {
      const form = wrapper.find(ServiceForm);
      expect(form.length).toEqual(0);
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.find(Button);
      expect(toggleButton.length).toEqual(1);
    })
  });

  describe("Form Holder state OPEN - NO Current Service Data",  () => {

    beforeAll(() => {
      window.scrollTo = jest.fn();
      wrapper = mount<{}, typeof ServiceFormHolder>(<ServiceFormHolder state={generateCleanState()} dispatch={jest.fn<React.Dispatch<ServiceAction>, []>()}/>)
    });

    it("Should Properly Mount Form Holder", () => {
      const toggleButton = wrapper.find("#serviceFormToggleBtn");
      toggleButton.at(0).simulate("click")
      // open button clicked //
      expect(wrapper.html()).toBeDefined();

    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#serviceFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have a Form Create Button", () => {
      const toggleButton = wrapper.render().find('#adminServiceFormCreate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered after toggle button", () => {
      const form = wrapper.find(ServiceForm);
      expect(form.length).toEqual(1);
    });
    it("Should NOT have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(ServiceImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(0);
    });
    it("Should NOT have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(ServiceImageUplForm);
      expect(imgUploadForm.length).toEqual(0);
    });

  });

  describe("Form Holder state OPEN - WITH Current Service Data - NO IMAGES",  () => {
    let state: IGlobalAppState;
    beforeAll(() => {
      window.scrollTo = jest.fn();
      state = setMockServiceState({ currentService: true });
      wrapper = mount<{}, typeof ServiceFormHolder>(<ServiceFormHolder state={state} dispatch={jest.fn<React.Dispatch<ServiceAction>, []>()}/>)
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.html()).toBeDefined();
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#serviceFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered", () => {
      const form = wrapper.find(ServiceForm);
      expect(form.length).toEqual(1);
    });
    it("Should have a Form Update Button", () => {
      const toggleButton = wrapper.render().find('#adminServiceFormUpdate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(ServiceImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(1);
    });
    it("Should NOT render any preview images", () => {
      const previewThumb = wrapper.find(ServiceImgPreviewThumb);
      expect(previewThumb.length).toEqual(0);
    })
    it("Should have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(ServiceImageUplForm);
      expect(imgUploadForm.length).toEqual(1);
    });

  });

  describe("Form Holder state OPEN - WITH Current Service Data - WITH IMAGES",  () => {
    let state: IGlobalAppState;
    beforeAll(() => {
      window.scrollTo = jest.fn();
      state = setMockServiceState({ currentService: true, serviceImages: 3 });
      wrapper = mount<{}, typeof ServiceFormHolder>(<ServiceFormHolder state={state} dispatch={jest.fn<React.Dispatch<ServiceAction>, []>()}/>)
    });

    it("Should Properly Mount Form Holder", () => {
      expect(wrapper.html()).toBeDefined();
    });
    it("Should have a Form toggle Button", () => {
      const toggleButton = wrapper.render().find('#serviceFormToggleBtn');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Form rendered", () => {
      const form = wrapper.find(ServiceForm);
      expect(form.length).toEqual(1);
    });
    it("Should have a Form Update Button", () => {
      const toggleButton = wrapper.render().find('#adminServiceFormUpdate');
      expect(toggleButton.length).toEqual(1);
    });
    it("Should have the Image Preview rendered", () => {
      const imgPreviewHolder = wrapper.find(ServiceImgPreviewHolder);
      expect(imgPreviewHolder.length).toEqual(1);
    });
    it("Should render a correct number of preview images", () => {
      const previewThumb = wrapper.find(ServiceImgPreviewThumb);
      const numberOfImages = state.serviceState.currentServiceData.images.length;
      expect(previewThumb.length).toEqual(numberOfImages);
    })
    it("Should have the Image Upload Form rendered", () => {
      const imgUploadForm = wrapper.find(ServiceImageUplForm);
      expect(imgUploadForm.length).toEqual(1);
    });
  });

});