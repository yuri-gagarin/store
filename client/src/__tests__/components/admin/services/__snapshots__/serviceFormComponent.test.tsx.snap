// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceFormHolder Component tests Default Form Holder state Should Properly Mount Form Holder 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "gqd4zb",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "gqd4zb",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(ServiceFormHolder)>
      <ServiceFormHolder
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "gqd4zb",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "gqd4zb",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "key": "gqd4zb",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
      >
        <div
          id="serviceFormHolder"
        >
          <GridRow>
            <div
              className="row"
            >
              <GridColumn
                computer={14}
                mobile={16}
                tablet={15}
              >
                <div
                  className="fourteen wide computer sixteen wide mobile fifteen wide tablet column"
                >
                  <Button
                    as="button"
                    content="Open Form"
                    id="serviceFormToggleBtn"
                    onClick={[Function]}
                  >
                    <Ref
                      innerRef={
                        Object {
                          "current": <button
                            class="ui button"
                            id="serviceFormToggleBtn"
                          >
                            Open Form
                          </button>,
                        }
                      }
                    >
                      <RefFindNode
                        innerRef={
                          Object {
                            "current": <button
                              class="ui button"
                              id="serviceFormToggleBtn"
                            >
                              Open Form
                            </button>,
                          }
                        }
                      >
                        <button
                          className="ui button"
                          id="serviceFormToggleBtn"
                          onClick={[Function]}
                        >
                          Open Form
                        </button>
                      </RefFindNode>
                    </Ref>
                  </Button>
                </div>
              </GridColumn>
            </div>
          </GridRow>
        </div>
      </ServiceFormHolder>
    </withRouter(ServiceFormHolder)>
  </Router>
</MemoryRouter>
`;

exports[`ServiceFormHolder Component tests Form Holder state OPEN - NO Current Service Data Should Properly Mount Form Holder, respond to '#serviceToggleBtn' click 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "xhlp0f",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "xhlp0f",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Component>
      <withRouter(ServiceFormHolder)>
        <ServiceFormHolder
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "key": "xhlp0f",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "key": "xhlp0f",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "key": "xhlp0f",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <div
            id="serviceFormHolder"
          >
            <GridRow>
              <div
                className="row"
              >
                <GridColumn
                  computer={14}
                  mobile={16}
                  tablet={15}
                >
                  <div
                    className="fourteen wide computer sixteen wide mobile fifteen wide tablet column"
                  >
                    <Button
                      as="button"
                      content="Close Form"
                      id="serviceFormToggleBtn"
                      onClick={[Function]}
                    >
                      <Ref
                        innerRef={
                          Object {
                            "current": <button
                              class="ui button"
                              id="serviceFormToggleBtn"
                            >
                              Close Form
                            </button>,
                          }
                        }
                      >
                        <RefFindNode
                          innerRef={
                            Object {
                              "current": <button
                                class="ui button"
                                id="serviceFormToggleBtn"
                              >
                                Close Form
                              </button>,
                            }
                          }
                        >
                          <button
                            className="ui button"
                            id="serviceFormToggleBtn"
                            onClick={[Function]}
                          >
                            Close Form
                          </button>
                        </RefFindNode>
                      </Ref>
                    </Button>
                    <ServiceForm
                      description=""
                      handleCreateService={[Function]}
                      handleUpdateService={[Function]}
                      name=""
                      price=""
                    >
                      <div
                        className="createServiceFormHolder"
                      >
                        <Form
                          as="form"
                          id="createServiceForm"
                        >
                          <form
                            className="ui form"
                            id="createServiceForm"
                            onSubmit={[Function]}
                          >
                            <FormField>
                              <div
                                className="field"
                              >
                                <label>
                                  Service name
                                </label>
                                <input
                                  onChange={[Function]}
                                  placeholder="Service name here ..."
                                  value=""
                                />
                              </div>
                            </FormField>
                            <FormField>
                              <div
                                className="field"
                              >
                                <label>
                                  Service price
                                </label>
                                <input
                                  onChange={[Function]}
                                  placeholder="Service price here..."
                                  value=""
                                />
                              </div>
                            </FormField>
                            <FormField
                              control={[Function]}
                              id="form-textarea-control-opinion"
                              label="Store Description"
                              onChange={[Function]}
                              placeholder="Service description here..."
                              value=""
                            >
                              <div
                                className="field"
                              >
                                <label
                                  htmlFor="form-textarea-control-opinion"
                                >
                                  Store Description
                                </label>
                                <TextArea
                                  aria-describedby={null}
                                  as="textarea"
                                  id="form-textarea-control-opinion"
                                  onChange={[Function]}
                                  placeholder="Service description here..."
                                  rows={3}
                                  value=""
                                >
                                  <Ref
                                    innerRef={
                                      Object {
                                        "current": <textarea
                                          id="form-textarea-control-opinion"
                                          placeholder="Service description here..."
                                          rows="3"
                                        />,
                                      }
                                    }
                                  >
                                    <RefFindNode
                                      innerRef={
                                        Object {
                                          "current": <textarea
                                            id="form-textarea-control-opinion"
                                            placeholder="Service description here..."
                                            rows="3"
                                          />,
                                        }
                                      }
                                    >
                                      <textarea
                                        aria-describedby={null}
                                        id="form-textarea-control-opinion"
                                        onChange={[Function]}
                                        onInput={[Function]}
                                        placeholder="Service description here..."
                                        rows={3}
                                        value=""
                                      />
                                    </RefFindNode>
                                  </Ref>
                                </TextArea>
                              </div>
                            </FormField>
                            <Button
                              as="button"
                              content="Create  New Service"
                              id="adminServiceFormCreate"
                              onClick={[Function]}
                              type="submit"
                            >
                              <Ref
                                innerRef={
                                  Object {
                                    "current": <button
                                      class="ui button"
                                      id="adminServiceFormCreate"
                                      type="submit"
                                    >
                                      Create  New Service
                                    </button>,
                                  }
                                }
                              >
                                <RefFindNode
                                  innerRef={
                                    Object {
                                      "current": <button
                                        class="ui button"
                                        id="adminServiceFormCreate"
                                        type="submit"
                                      >
                                        Create  New Service
                                      </button>,
                                    }
                                  }
                                >
                                  <button
                                    className="ui button"
                                    id="adminServiceFormCreate"
                                    onClick={[Function]}
                                    type="submit"
                                  >
                                    Create  New Service
                                  </button>
                                </RefFindNode>
                              </Ref>
                            </Button>
                          </form>
                        </Form>
                      </div>
                    </ServiceForm>
                  </div>
                </GridColumn>
              </div>
            </GridRow>
          </div>
        </ServiceFormHolder>
      </withRouter(ServiceFormHolder)>
    </Component>
  </Router>
</MemoryRouter>
`;
